docs/development/architecture.md

```diff
@@ -0,0 +1,221 @@
+---
+title: Architecture
+sidebar_position: 0
+---
+
+
+## Introduction
+
+This document explains how Headlamp is built. It covers the main parts of the system, how you can run it, how to customize it, and the ideas behind its design.
+
+Headlamp is a user-friendly and extensible Kubernetes UI. Designed to provide a seamless experience for managing Kubernetes clusters. It can connect to one or multiple Kubernetes clusters and can be deployed as a web app in-cluster or as a desktop app. You can customize Headlamp using plugins, which let you change how it looks and works to fit your needs.
+
+## High-Level Architecture and Request Lifecycle
+
+The diagrams below show how Headlamp works and the different ways you can run it.
+
+### Data Flow / Request Lifecycle
+
+How does a typical user request flow through Headlamp?
+
+#### Headlamp running as a Desktop App (on MacOS, Windows, Linux etc)
+
+```mermaid
+graph TD
+    A[User] -->|Interacts with| B[Headlamp App]
+    B -->|Runs| C[Frontend, React]
+    B -->|Runs| D[Backend headlamp-server, Go]
+    D -->|Proxies| E[Kubernetes API]
+    C -->|Loads| F[Plugins]
+    D -->|Serves static web files and provides services to| C
+```
+
+#### Headlamp Web App running inside a Kubernetes Cluster
+
+```mermaid
+graph TD
+    A[User] -->|Interacts with| G[Ingress/Gateway API]
+    G -->|Routes to| B[Frontend, React]
+    B -->|Sends requests to| C[Backend headlamp-server, Go]
+    C -->|Proxies| D[Kubernetes API]
+    C -->|Interacts with| E[Cluster Services]
+    B -->|Loads| F[Plugins]
+    C -->|Serves static web files and provides services to| B
+    subgraph Kubernetes Cluster
+        B
+        C
+        E
+    end
+```
+
+Notes:
+- Frontend handles user interactions and displays data.
+- Backend acts as a proxy and service layer. Handling authentication, RBAC, and API aggregation.
+- Kubernetes API is the source of truth for cluster state.
+- WebSocket connections follow a similar path. However they maintain persistent connections for real-time updates.
+
+
+### WebSocket Lifecycle and Real-Time Updates
+
+How does headlamp get real-time updates from Kubernetes clusters?
+
+- Websockets are used for real-time updates from Kubernetes. 
+- Browsers support a limited number of connections. Six usually.
+- With Headlamp in the browser there is one websocket to `headlamp-server` backend. The `headlamp-server` connects to many websockets on different Kubernetes API servers. Overcoming the browser limit.
+
+The diagram below illustrates how Headlamp manages WebSocket connections between the browser, the `headlamp-server`, and multiple Kubernetes API servers:
+
+```mermaid
+graph LR
+    A[Browser] <-->|WebSocket| B[headlamp_server]
+    B <-->|WebSocket| C[Kubernetes API Server]
+    B <-->|WebSocket| D[Kubernetes API Server]
+    B <-->|WebSocket| E[Kubernetes API Server]
+```
+
+### Plugins
+
+#### Plugin Hosting Options
+
+Where can plugins be hosted?
+
+```mermaid
+graph TD
```

Comment:
![image](https://github.com/user-attachments/assets/3c4967f2-0676-4d78-8686-a29955861565)

Edges are duplicated in this graph, also does graph make sense here? Just a list with two items would work fine imho

Reply:
Made some improvements to this section.